packages/flutter_tools/lib/src/commands/build_macos_framework.dart

```diff
@@ -0,0 +1,305 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../base/common.dart';
+import '../base/file_system.dart';
+import '../base/logger.dart';
+import '../base/process.dart';
+import '../base/utils.dart';
+import '../build_info.dart';
+import '../build_system/build_system.dart';
+import '../build_system/targets/macos.dart';
+import '../cache.dart';
+import '../flutter_plugins.dart';
+import '../globals.dart' as globals;
+import '../macos/cocoapod_utils.dart';
+import '../runner/flutter_command.dart' show DevelopmentArtifact, FlutterCommandResult;
+import '../version.dart';
+import 'build_ios_framework.dart';
+
+/// Produces a .framework for integration into a host macOS app. The .framework
+/// contains the Flutter engine and framework code as well as plugins. It can
+/// be integrated into plain Xcode projects without using or other package
+/// managers.
+class BuildMacOSFrameworkCommand extends BuildFrameworkCommand {
+  BuildMacOSFrameworkCommand({
+    super.flutterVersion,
+    required super.buildSystem,
+    required super.verboseHelp,
+    super.cache,
+    super.platform,
+  });
+
+  @override
+  final String name = 'macos-framework';
+
+  @override
+  final String description = 'Produces .xcframeworks for a Flutter project '
+      'and its plugins for integration into existing, plain macOS Xcode projects.\n'
+      'This can only be run on macOS hosts.';
+
+  @override
+  Future<Set<DevelopmentArtifact>> get requiredArtifacts async => const <DevelopmentArtifact>{
+        DevelopmentArtifact.macOS,
+      };
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    final String outputArgument =
+        stringArg('output') ?? globals.fs.path.join(globals.fs.currentDirectory.path, 'build', 'macos', 'framework');
+
+    if (outputArgument.isEmpty) {
+      throwToolExit('--output is required.');
+    }
+
+    if (!project.macos.existsSync()) {
+      throwToolExit('Project does not support macOS');
+    }
+
+    final Directory outputDirectory =
+        globals.fs.directory(globals.fs.path.absolute(globals.fs.path.normalize(outputArgument)));
+    final List<BuildInfo> buildInfos = await getBuildInfos();
+    displayNullSafetyMode(buildInfos.first);
+    for (final BuildInfo buildInfo in buildInfos) {
+      globals.printStatus('Building macOS frameworks in ${getNameForBuildMode(buildInfo.mode)} mode...');
+      final String xcodeBuildConfiguration = sentenceCase(getNameForBuildMode(buildInfo.mode));
+      final Directory modeDirectory = outputDirectory.childDirectory(xcodeBuildConfiguration);
+
+      if (modeDirectory.existsSync()) {
+        modeDirectory.deleteSync(recursive: true);
+      }
+
+      if (boolArg('cocoapods') ?? false) {
```

Comment:
I'm guessing the answer is "no", but should we make these `boolArg(...)!`? (I believe) the only way this would error out is if a later refactor to the arg parsing results in this command no longer having a `cocoapods` flag and this code wasn't updated, which would be a tool bug that we'd want to know about.

Reply:
@Jasguerrero I keep running into this pattern.  Can we make `boolArg` return a nonnull bool to simplify the code, and have it throw an exception (not `toolExit` since it's a developer error) if the flag doesn't exist?

Reply:
Filed https://github.com/flutter/flutter/issues/105263 to track